lib/logstash/outputs/tcp.rb

```diff
@@ -51,31 +51,45 @@ class LogStash::Outputs::Tcp < LogStash::Outputs::Base
   # SSL key passphrase
   config :ssl_key_passphrase, :validate => :password, :default => nil
 
+  ##
+  # @param socket [Socket]
+  # @param logger_context [#log_warn&#log_error&#logger]
   class Client
-    public
-    def initialize(socket, logger)
+    def initialize(socket, logger_context)
       @socket = socket
-      @logger = logger
+      @peer_info = socket.peer
+      @logger_context = logger_context
       @queue  = Queue.new
     end
 
-    public
     def run
       loop do
         begin
-          @socket.write(@queue.pop)
+          payload = @queue.pop
+
+          @logger_context.logger.trace("transmitting #{payload.bytesize} bytes", socket: @peer_info) if @logger_context.logger.trace? && payload && !payload.empty?
+          while payload && !payload.empty?
+            written_bytes_size = @socket.write(payload)
+            payload = payload.byteslice(written_bytes_size..-1)
+            @logger_context.logger.log_trace(">transmitted #{written_bytes_size} bytes; #{payload.bytesize} bytes remain", socket: @peer_info) if @logger_context.logger.trace?
+            sleep 0.1 unless payload.empty?
```

Comment:
It's not clean to me why we need to sleep 10ms if the `payload` contains data. This is a rate limiting at the end, which would suffer the performance every time we have payloads bigger then the window.
I think that it the `socket.write` discover it doesn't have free space in network stack buffer, it blocks or return an error.
If it returns an error we don't recover the sending.
If it blocks, we don't have problems and shouldn't wait later.